specs/dataclass_transforms.txt

```diff
@@ -0,0 +1,399 @@
+    # describe default value
+    office_number = "unassigned"
+```
+
+Libraries that support dataclass-like semantics and support field descriptor
+classes typically use common parameter names to construct these field
+descriptors. This specification formalizes the names and meanings of the
+parameters that must be understood for static type checkers. These standardized
+parameters must be keyword-only parameters. Field descriptor classes are
+allowed to use other parameters in their constructors, and those parameters
+can be positional and may use other names.
+
+
+`init` is an optional bool parameter that indicates whether the field should
+be included in the synthesized `__init__` method. If unspecified, it defaults
+to True.
+
+`default` is an optional parameter that provides the default value for the
+field.
+
+`default_factory` is an optional parameter that provides a runtime callback
+that returns the default value for the field. If `default` and `default_value`
+are both unspecified, the field is assumed to have no default value and must be
+provided a value when the class is instantiated.
+
+`alias` is an optional str parameter that provides an alternative name for
+the field. This alternative name is used in the synthesized `__init__` method.
+
+
+This example demonstrates 
+```python
+# Library code (within type stub or inline):
+class ModelField:
+    def __init__(
+        self,
+        *,
+        default: Optional[Any] = ...,
+        init: Optional[bool] = True,
+        **kwargs: Any
+    ) -> None: ...
+
+@typing.dataclass_transform(kw_only_default=True, field_descriptors=(ModelField, ))
+def create_model(cls: type) -> Callable[[_T], _T]: ...
+
+
+# Code that imports this library:
+@create_model(init=False)
+class Customer:
+    id: int = ModelField(default=0)
+    name: str
+```
+
+
+Runtime Behavior
+----------------
+
+At runtime, the `dataclass_transform` decorator has no effect. It simply returns
+a function that accepts a single argument and returns that argument as the
+return value.
+
+Here is its complete implementation.
+
+```python
+def dataclass_transform(
+    *,
+    eq_default: bool = True,
+    order_default: bool = False,
+    kw_only_default: bool = False,
+    field_descriptors: Tuple[Union[type, Callable[..., Any]], ...] = (()),
+) -> Callable[[_T], _T]:
+    return lambda a: a
+```
+
+
+Dataclass Semantics
+-------------------
+
+The following dataclass semantics are implied when dataclass_transform is
+specified.
+
+Frozen classes cannot inherit from non-frozen classes. A class that
+directly specifies a metaclass that has been decorated with
+`dataclass_transform` will not be considered non-frozen. In the example
+
+```python
+@typing.dataclass_transform()
+class ModelMeta(type): ...
+
+# ModelBase is not considered either "frozen" or "non-frozen"
+# because it directly specifies ModelMeta as its metaclass.
+class ModelBase(metaclass=ModelMeta): ...
```

Comment:
so here type checker treats ModelMeta as dataclass ? and apply dataclass rules? and this, frozen, non-frozen behavior, is normal data class behavior?